ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,721 @@
+	return footer
+}
+
+func decodeTopLevelNodesAndTriesFooter(footer []byte) (uint64, uint16, error) {
+	const footerSize = encNodeCountSize + encTrieCountSize // footer doesn't include crc32 sum
+	if len(footer) != footerSize {
+		return 0, 0, fmt.Errorf("wrong footer size, expect %v, got %v", footerSize, len(footer))
+	}
+	nodesCount := binary.BigEndian.Uint64(footer)
+	triesCount := binary.BigEndian.Uint16(footer[encNodeCountSize:])
+	return nodesCount, triesCount, nil
+}
+
+func encodeNodeCount(nodeCount uint64) []byte {
+	footer := make([]byte, encNodeCountSize)
+	binary.BigEndian.PutUint64(footer, nodeCount)
+	return footer
+}
+
+func decodeNodeCount(encoded []byte) (uint64, error) {
+	if len(encoded) != encNodeCountSize {
+		return 0, fmt.Errorf("wrong subtrie node count size, expect %v, got %v", encNodeCountSize, len(encoded))
+	}
+	return binary.BigEndian.Uint64(encoded), nil
+}
+
+func encodeCRC32Sum(checksum uint32) []byte {
+	buf := make([]byte, crc32SumSize)
+	binary.BigEndian.PutUint32(buf, checksum)
+	return buf
+}
+
+func decodeCRC32Sum(encoded []byte) (uint32, error) {
+	if len(encoded) != crc32SumSize {
+		return 0, fmt.Errorf("wrong crc32sum size, expect %v, got %v", crc32SumSize, len(encoded))
+	}
+	return binary.BigEndian.Uint32(encoded), nil
+}
+
+func encodeVersion(magic uint16, version uint16) []byte {
+	// Write header: magic (2 bytes) + version (2 bytes)
+	header := make([]byte, encMagicSize+encVersionSize)
+	binary.BigEndian.PutUint16(header, magic)
+	binary.BigEndian.PutUint16(header[encMagicSize:], version)
+	return header
+}
+
+func decodeVersion(encoded []byte) (uint16, uint16, error) {
+	if len(encoded) != encMagicSize+encVersionSize {
+		return 0, 0, fmt.Errorf("wrong version size, expect %v, got %v", encMagicSize+encVersionSize, len(encoded))
+	}
+	magicBytes := binary.BigEndian.Uint16(encoded)
+	version := binary.BigEndian.Uint16(encoded[encMagicSize:])
+	return magicBytes, version, nil
+}
+
+func encodeSubtrieCount(level uint16) []byte {
+	bytes := make([]byte, encSubtrieCountSize)
+	binary.BigEndian.PutUint16(bytes, level)
+	return bytes
+}
+
+func decodeSubtrieCount(encoded []byte) (uint16, error) {
+	if len(encoded) != encSubtrieCountSize {
+		return 0, fmt.Errorf("wrong subtrie level size, expect %v, got %v", encSubtrieCountSize, len(encoded))
+	}
+	return binary.BigEndian.Uint16(encoded), nil
+}
+
+// closeAndMergeError close the closable and merge the closeErr with the given err into a multierror
+// Note: when using this function in a defer function, don't use as below:
+// func XXX() (
+//
+//	err error,
+//	) {
+//		def func() {
+//			// bad, because the definition of err might get overwritten
+//			err = closeAndMergeError(closable, err)
+//		}()
+//
+// Better to use as below:
+// func XXX() (
+//
+//	errToReturn error,
+//	) {
+//		def func() {
+//			// good, because the error to returned is only updated here, and guaranteed to be returned
+//			errToReturn = closeAndMergeError(closable, err)
```

Comment:
Need to use `errToReturn` as function argument. 

```suggestion
//			errToReturn = closeAndMergeError(closable, errToReturn)
```